From f035b1fcded7851fdcd1f95edb3b0d6020614bb0 Mon Sep 17 00:00:00 2001
From: Cliff Maceyak <cliffmaceyak@gmail.com>
Date: Thu, 4 Nov 2021 00:00:00 +0000
Subject: [PATCH] Forbid dynamic code generation on Linux

The Windows sandbox sets MITIGATION_DYNAMIC_CODE_DISABLE on certain
processes that do not generate dynamic code; we forbid dynamic code
generation in the same processes via seccomp.

Signed-off-by: Cliff Maceyak <cliffmaceyak@gmail.com>
---
 .../policy/linux/bpf_audio_policy_linux.cc    | 18 ++++++++++++++
 .../policy/linux/bpf_ppapi_policy_linux.cc    | 12 ++++++++++
 .../policy/linux/bpf_renderer_policy_linux.cc | 21 +++++++++++++++-
 .../policy/linux/bpf_renderer_policy_linux.h  |  5 +++-
 .../policy/linux/sandbox_seccomp_bpf_linux.cc | 24 ++++++++++++++++++-
 5 files changed, 77 insertions(+), 3 deletions(-)

diff --git a/sandbox/policy/linux/bpf_audio_policy_linux.cc b/sandbox/policy/linux/bpf_audio_policy_linux.cc
index 1ca63de3862b6..358839dbfc950 100644
--- a/sandbox/policy/linux/bpf_audio_policy_linux.cc
+++ b/sandbox/policy/linux/bpf_audio_policy_linux.cc
@@ -118,6 +118,24 @@ ResultExpr AudioProcessPolicy::EvaluateSyscall(int system_call_number) const {
       const Arg<int> domain(0);
       return If(domain == AF_UNIX, Allow()).Else(Error(EPERM));
     }
+#endif
+#if defined(__i386__) || defined(__x86_64__) || defined(__mips__) || \
+    defined(__aarch64__)
+    case __NR_mmap:
+#endif
+#if defined(__i386__) || defined(__arm__) || \
+    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
+    case __NR_mmap2:
+#endif
+      return RestrictMmapFlagsNoWX();
+    case __NR_mprotect:
+    case __NR_pkey_mprotect:
+      return RestrictMprotectFlagsNoWX();
+#if defined(__i386__) || defined(__x86_64__) || defined(__arm__) || \
+    defined(__aarch64__) ||                                         \
+    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_64_BITS))
+    case __NR_shmat:
+      return RestrictShmatFlags();
 #endif
     default:
 #if defined(__x86_64__)
diff --git a/sandbox/policy/linux/bpf_ppapi_policy_linux.cc b/sandbox/policy/linux/bpf_ppapi_policy_linux.cc
index 1465de67d7a50..ae3d240df3363 100644
--- a/sandbox/policy/linux/bpf_ppapi_policy_linux.cc
+++ b/sandbox/policy/linux/bpf_ppapi_policy_linux.cc
@@ -40,6 +40,18 @@ ResultExpr PpapiProcessPolicy::EvaluateSyscall(int sysno) const {
       return RestrictSchedTarget(GetPolicyPid(), sysno);
     case __NR_ioctl:
       return Error(ENOTTY);  // Flash Access.
+#if defined(__i386__) || defined(__x86_64__) || defined(__mips__) || \
+    defined(__aarch64__)
+    case __NR_mmap:
+#endif
+#if defined(__i386__) || defined(__arm__) || \
+    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
+    case __NR_mmap2:
+#endif
+      return RestrictMmapFlagsNoWX();
+    case __NR_mprotect:
+    case __NR_pkey_mprotect:
+      return RestrictMprotectFlagsNoWX();
     default:
       // Default on the baseline policy.
       return BPFBasePolicy::EvaluateSyscall(sysno);
diff --git a/sandbox/policy/linux/bpf_renderer_policy_linux.cc b/sandbox/policy/linux/bpf_renderer_policy_linux.cc
index f789e92c37c1b..1bc2d82e5098b 100644
--- a/sandbox/policy/linux/bpf_renderer_policy_linux.cc
+++ b/sandbox/policy/linux/bpf_renderer_policy_linux.cc
@@ -48,10 +48,29 @@ ResultExpr RestrictIoctl() {
 
 }  // namespace
 
-RendererProcessPolicy::RendererProcessPolicy() {}
+RendererProcessPolicy::RendererProcessPolicy(bool is_jit_disabled)
+    : is_jit_disabled_(is_jit_disabled) {}
 RendererProcessPolicy::~RendererProcessPolicy() {}
 
 ResultExpr RendererProcessPolicy::EvaluateSyscall(int sysno) const {
+
+  if (is_jit_disabled_) {
+    switch (sysno) {
+#if defined(__i386__) || defined(__x86_64__) || defined(__mips__) || \
+    defined(__aarch64__)
+      case __NR_mmap:
+#endif
+#if defined(__i386__) || defined(__arm__) || \
+    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
+      case __NR_mmap2:
+#endif
+        return RestrictMmapFlagsNoWX();
+      case __NR_mprotect:
+      case __NR_pkey_mprotect:
+        return RestrictMprotectFlagsNoWX();
+    }
+  }
+
   switch (sysno) {
     // The baseline policy allows __NR_clock_gettime. Allow
     // clock_getres() for V8. crbug.com/329053.
diff --git a/sandbox/policy/linux/bpf_renderer_policy_linux.h b/sandbox/policy/linux/bpf_renderer_policy_linux.h
index c05e3043d47d4..f2a7a5b256528 100644
--- a/sandbox/policy/linux/bpf_renderer_policy_linux.h
+++ b/sandbox/policy/linux/bpf_renderer_policy_linux.h
@@ -14,7 +14,7 @@ namespace policy {
 // This policy can be used by both renderer and worker processes.
 class RendererProcessPolicy : public BPFBasePolicy {
  public:
-  RendererProcessPolicy();
+  explicit RendererProcessPolicy(bool is_jit_disabled);
 
   RendererProcessPolicy(const RendererProcessPolicy&) = delete;
   RendererProcessPolicy& operator=(const RendererProcessPolicy&) = delete;
@@ -22,6 +22,9 @@ class RendererProcessPolicy : public BPFBasePolicy {
   ~RendererProcessPolicy() override;
 
   bpf_dsl::ResultExpr EvaluateSyscall(int system_call_number) const override;
+
+ private:
+  const bool is_jit_disabled_;  // Disable dynamic code generation if jitless
 };
 
 }  // namespace policy
diff --git a/sandbox/policy/linux/sandbox_seccomp_bpf_linux.cc b/sandbox/policy/linux/sandbox_seccomp_bpf_linux.cc
index 4c8725e86a926..2fdf4166d0db7 100644
--- a/sandbox/policy/linux/sandbox_seccomp_bpf_linux.cc
+++ b/sandbox/policy/linux/sandbox_seccomp_bpf_linux.cc
@@ -19,6 +19,7 @@
 #include "base/notreached.h"
 #include "build/build_config.h"
 #include "build/chromeos_buildflags.h"
+#include "gin/gin_features.h"
 #include "printing/buildflags/buildflags.h"
 #include "sandbox/linux/bpf_dsl/bpf_dsl.h"
 #include "sandbox/linux/bpf_dsl/trap_registry.h"
@@ -50,6 +51,7 @@
 #include "sandbox/policy/linux/bpf_service_policy_linux.h"
 #include "sandbox/policy/linux/bpf_speech_recognition_policy_linux.h"
 #include "sandbox/policy/linux/bpf_utility_policy_linux.h"
+#include "third_party/blink/common/switches.h"
 
 #if !defined(OS_NACL_NONSFI)
 #include "sandbox/policy/chromecast_sandbox_allowlist_buildflags.h"
@@ -173,7 +175,27 @@ std::unique_ptr<BPFBasePolicy> SandboxSeccompBPF::PolicyForSandboxType(
     case SandboxType::kGpu:
       return GetGpuProcessSandbox(options.use_amd_specific_policies);
     case SandboxType::kRenderer:
-      return std::make_unique<RendererProcessPolicy>();
+      const base::CommandLine& command_line =
+          *base::CommandLine::ForCurrentProcess();
+      bool dynamic_code_can_be_disabled = false;
+      if (base::FeatureList::IsEnabled(features::kV8NoJIT)) {
+        dynamic_code_can_be_disabled = true;
+      }
+      else if (command_line->HasSwitch(blink::switches::kJavaScriptFlags)) {
+        std::string js_flags =
+            command_line->GetSwitchValueASCII(blink::switches::kJavaScriptFlags);
+        std::vector<base::StringPiece> js_flag_list = base::SplitStringPiece(
+            js_flags, ",", base::TRIM_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+        for (const auto& js_flag : js_flag_list) {
+          if (js_flag == "--jitless") {
+            // If v8 is running jitless then there is no need for the ability to
+            // mark writable pages as executable to be available to the process.
+            dynamic_code_can_be_disabled = true;
+            break;
+          }
+        }
+      }
+      return std::make_unique<RendererProcessPolicy>(dynamic_code_can_be_disabled);
 #if BUILDFLAG(ENABLE_PLUGINS)
     case SandboxType::kPpapi:
       return std::make_unique<PpapiProcessPolicy>();
-- 
2.30.1 (Apple Git-130)

